app/controllers/api/v1/scores_controller.rb

```diff
@@ -0,0 +1,120 @@
+class Api::V1::ScoresController < Api::V1::BaseApiController
+
+  before_action -> { require_privilege :instructor_all }
+  before_action :set_assessment
+
+  def index
+    submissions = @assessment.submissions.where(assessment_id: @assessment.id)
+
+    scores = {}
+    cud_to_user = {}
+    submissions.each do |submission|
+      cud = submission.course_user_datum_id
+      unless cud_to_user.key?(cud)
+        cud_to_user[cud] = User.find(CourseUserDatum.find(cud).user_id)
+      end
+      user = cud_to_user[cud]
+
+      raw_submission_scores = Score.where(submission_id: submission.id)
+      user_scores = scores.fetch(user.email, {})
+      user_scores[submission.version] = problem_name_to_score(raw_submission_scores)
+
+      scores[user.email] = user_scores
+    end
+
+    respond_with scores
+  end
+
+  def show
+    cud = get_cud params
+    submissions = @assessment.submissions.where(
+      assessment_id: @assessment.id,
+      course_user_datum_id: cud.id
+    )
+
+    scores = {}
+    submissions.each do |submission|
+      raw_submission_scores = Score.where(submission_id: submission.id)
+      scores[submission.version] = problem_name_to_score(raw_submission_scores)
+    end
+
+    respond_with scores
+  end
+
+  def update
+    cud = get_cud params
+    aud = AssessmentUserDatum.find_by(assessment_id: @assessment.id, course_user_datum_id: cud.id)
+
+    cuds = []
+    update_group_scores = params.key?(:update_group_scores) &&
+                          ActiveModel::Type::Boolean.new.cast(params[:update_group_scores].downcase)
+
+    if aud.group_id.nil? || !update_group_scores
+      cuds.append(cud.id)
+    else
+      auds = AssessmentUserDatum.where(assessment_id: @assessment.id, group_id: aud.group_id)
+      auds.each do |aud|
+        cuds.append(aud.course_user_datum_id)
+      end
+    end
+
+    scores = {}
+    cuds.each do |cud_id|
+      submission = @assessment.submissions.where(
+        assessment_id: @assessment.id,
+        course_user_datum_id: cud_id
+      ).order(version: :desc).first
+
+      problem_id_to_name = @assessment.problem_id_to_name
+      submission_scores = Score.where(submission_id: submission.id)
+
+      user_scores = {}
+      submission_scores.each do |score|
+        problem_name = problem_id_to_name[score.problem_id]
+        user_scores[problem_name] = score.score
+        next unless params.key?(problem_name)
+
+        score.score = params[problem_name]
+        unless score.save
```

Comment:
In this case, it's not good to save the scores within this loop. Because if there's an error, the previous score would already have been saved.

For example, you want to update user 1, 2, 3 & 4's problem 1 score to 100.
user1 and user2 saves, but user3 errors out. So now user1 & user2 score has been changed, but user3 hasn't been changed, and this makes the endpoint's "Error case" have side effects. In general, if an error occurs in a batch update, no changes should be made.

Reply:
What I recommend is to keep track of all the scores to be saved in a list, and then use a activerecord transaction to save all of them. If one of it fails, then it rollsback all of them

https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html

Reply:
Will change to use `Score.transaction`